popde/adaptive_kde.py

```diff
@@ -224,6 +224,151 @@ def optimize_parameters(self, method='loo_cv', fom_plot_name=None):
+            do_fit (bool, optional): Whether to fit the KDE upon initialization.
+            n_splits (int, optional): Number of splits for k-fold 
+               cross-validation.
+        """
+
+        self.n_splits = n_splits
+        self.alpha = alpha
+        self.bandwidth = bandwidth
+
+        super().__init__(data, weights, input_transf, stdize, rescale, backend, 
+                bandwidth, alpha, dim_names, do_fit)
+
+    def loo_cv_score(self, rescale_factors_alpha):
+        """
+        Compute the Leave-One-Out Cross-Validation (LOO-CV) score
+          based on log-likelihood.
+
+        Args:
+            rescale_factors_alpha (array-like): Array of rescale factors per
+              dimension and alpha value.
+
+        Returns:
+            float: Negative sum of log-likelihood scores for LOO-CV.
+        """
+        rescale_factors = rescale_factors_alpha[:-1]
+        alpha_choice = rescale_factors_alpha[-1]
+        from sklearn.model_selection import LeaveOneOut
+        loo = LeaveOneOut() 
+        fom = 0.
+        for train_index, test_index in loo.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None # FIX ME
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factors,
+                                  bandwidth=self.bandwidth, alpha=alpha_choice)
+            fom += np.log(awkde.evaluate_with_transf(test_data))
+        return -fom
+
+    def kfold_cv_score(self, rescale_factors_alpha, seed=42):
+        """
+        Perform k-fold cross-validation
+        """
+        rescale_factors = rescale_factors_alpha[:-1]
+        alpha_choice = rescale_factors_alpha[-1]
+        from sklearn.model_selection import KFold
+        kf = KFold(n_splits=self.n_splits, shuffle=True, random_state=seed)
+        fom = []
+        for train_index, test_index in kf.split(self.data):
+            train_data, test_data = self.data[train_index], self.data[test_index]
+            local_weights = None # FIX ME
+
+            awkde = AdaptiveBwKDE(train_data, local_weights, input_transf=self.input_transf,
+                                  stdize=self.stdize, rescale=rescale_factors,
+                                  bandwidth=self.bandwidth, alpha=alpha_choice, do_fit=True)
+
+            log_kde_eval = np.log(awkde.evaluate_with_transf(test_data))
+            fom.append(log_kde_eval.sum())
+        return -sum(fom)
+
+    def optimize_rescale_parameters(self, initial_rescale_factor, initial_alpha=0.0, method='kfold_cv', bounds=None):
+
+        """
+        given initial choice of rescale factors in each dimension
+        and alpha perform optimization based on 
+        cross validations on data
+        """
+        #make alpha bounds to be fixed in [0, 1]
+        if bounds is None:
```

Comment:
I think it's OK to supply some default bounds like (0.01, 100) in each dimension as it's unlikely that the rescaling would go outside these.

Reply:
done